exporters/otlp/otlpgrpc/connection.go

```diff
@@ -276,3 +282,127 @@ func (c *connection) contextWithStop(ctx context.Context) (context.Context, cont
 	}(ctx, cancel)
 	return ctx, cancel
 }
+
+func (c *connection) doRequest(ctx context.Context, fn func(context.Context) error) error {
+	expBackoff := newExponentialBackoff(c.cfg.RetrySettings)
+
+	for {
+		err := fn(ctx)
+		if err == nil {
+			// request succeeded.
+			return nil
+		}
+
+		if !c.cfg.RetrySettings.Enabled {
+			return err
+		}
+
+		// We have an error, check gRPC status code.
+		st := status.Convert(err)
+		if st.Code() == codes.OK {
+			// Not really an error, still success.
+			return nil
+		}
+
+		// Now, this is this a real error.
+
+		if !shouldRetry(st.Code()) {
+			// It is not a retryable error, we should not retry.
+			return err
+		}
+
+		// Need to retry.
+		var delay time.Duration
+
+		// Respect server throttling.
+		throttle := getThrottleDuration(st)
+
+		backoffDelay := expBackoff.NextBackOff()
+		if backoffDelay == backoff.Stop {
+			// throw away the batch
+			err = fmt.Errorf("max elapsed time expired %w", err)
+			return err
+		}
+
+		if backoffDelay > throttle {
+			delay = backoffDelay
+		} else {
+			delay = throttle
+		}
+
+		// back-off, but get interrupted when shutting down or request is cancelled or timed out.
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case <-c.stopCh:
+			return fmt.Errorf("interrupted due to shutdown %w", err)
+		case <-time.After(delay):
+		}
+	}
+}
+
+func shouldRetry(code codes.Code) bool {
+	switch code {
+	case codes.OK:
+		// Success. This function should not be called for this code, the best we
+		// can do is tell the caller not to retry.
+		return false
+
+	case codes.Canceled,
+		codes.DeadlineExceeded,
+		codes.PermissionDenied,
+		codes.Unauthenticated,
+		codes.ResourceExhausted,
+		codes.Aborted,
+		codes.OutOfRange,
+		codes.Unavailable,
+		codes.DataLoss:
+		// These are retryable errors.
+		return true
```

Comment:
`PermissionDenied` and `Unauthenticated` seem are not retryable errors. Whatever we retry, it keeps failing without fixing the authorization problem.

Reply:
Hm, the specs define those errors as retryable, but indeed that is strange. The oteps linked in the specs define both as not retryable as you suggest. 

I think we should move forward with the way the specs define it today. I'll ask specs-sig if those errors are really supposed to be retryable or not, if not I'll update this and the specs.

Reply:
I think those are both typically viewed as retryable errors as they're not necessarily permanent failures and the situation may have changed on a subsequent request.  It's unlikely with `Unauthenticated` that making the exact same request will suddenly work, but the authentication requirement could change or be removed.  Similarly, `PermissionDenied` will likely not go away absent operator intervention, but it is plausible to think that authorization had been granted but that information hadn't yet propagated to the AuthZ decision-maker on the receiving side and it might have done so on a subsequent request. v0v

Reply:
I get that case @Aneurysm9 just not sure if we should handle that. I was double checking the specs, and there is a deviation if these should be retryable or not. 
- https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/protocol/otlp.md#otlpgrpc-response
- https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/protocol/exporter.md#transient-errors


I'm going to follow the first link that says that those errors are *not* retryable since it comes from an OTEP that the second link was based on. Also, I'll open a PR updating the specs to fix that.

Reply:
I've just opened the [PR in the specs repo](https://github.com/open-telemetry/opentelemetry-specification/pull/1642).